factom/client.py

```diff
@@ -301,31 +301,58 @@ def transaction(self, tx_hash: Union[bytes, str]):
 
     # Convenience methods
 
-    def read_chain(self, chain_id: Union[bytes, str], include_entry_context: bool = False, encode_as_hex: bool = False):
-        """
-        Shortcut method to read an entire chain.
-
-        Args:
-            chain_id (str): Chain ID to read.
-            include_entry_context (bool): Whether to include information like entry hash, timestamp, and block height
-            encode_as_hex (bool): Whether to encode external-ids and content in entry objects as hex or not
-
-        Returns:
-            list[dict]: A list of entry dictionaries in reverse chronologial order.
-        """
-        entries = []
+    def read_chain(
+        self,
+        chain_id: Union[bytes, str],
+        from_height: int = 0,
+        include_entry_context: bool = False,
+        encode_as_hex: bool = False,
+    ):
+        """A generator that yields all entries of a chain in order, optionally starting from a given block height."""
+        # Walk the entry block chain backwards to build up a stack of entry blocks to fetch
+        entry_blocks = []
         keymr = self.chain_head(chain_id)["chainhead"]
         while keymr != NULL_BLOCK:
             block = self.entry_block(keymr)
-            for entry_pointer in reversed(block["entrylist"]):
+            if block["header"]["dbheight"] < from_height:
+                break
+            entry_blocks.append(block)
+            keymr = block["header"]["prevkeymr"]
+
+        # Continuously pop off the stack and yield each entry one by one (in the order that they appear in the block)
+        while len(entry_blocks) > 0:
+            entry_block = entry_blocks.pop()
+            for entry_pointer in entry_block["entrylist"]:
                 entry = self.entry(entry_pointer["entryhash"], encode_as_hex=encode_as_hex)
                 if include_entry_context:
                     entry["entryhash"] = entry_pointer["entryhash"]
                     entry["timestamp"] = entry_pointer["timestamp"]
-                    entry["dbheight"] = block["header"]["dbheight"]
-                entries.append(entry)
-            keymr = block["header"]["prevkeymr"]
-        return entries
+                    entry["dbheight"] = entry_block["header"]["dbheight"]
+                yield entry
```

Comment:
I'm wondering if it wouldn't be better to just switch to a completely lazy implementation (getting rid of the accumulation of the `entry_blocks`). Wouldn't something like this work?

```python
keymr = self.chain_head(chain_id)["chainhead"]
while keymr != NULL_BLOCK:
    entry_block = self.entry_block(keymr)
    if entry_block["header"]["dbheight"] < from_height:
        continue
    for entry_pointer in entry_block["entrylist"]:
        entry = self.entry(entry_pointer["entryhash"], encode_as_hex=encode_as_hex)
        if include_entry_context:
            entry["entryhash"] = entry_pointer["entryhash"]
            entry["timestamp"] = entry_pointer["timestamp"]
            entry["dbheight"] = entry_block["header"]["dbheight"]
        yield entry
    keymr = block["header"]["prevkeymr"]
```

Reply:
This one's kinda related to the previous comment as well. I don't think we can skip the accumulation of entry blocks, since we have to walk backwards in order to get to the first entry in the requested sequence. In your code snippet, it seems like it would yield all entries in latest block, then all in the one before that, and so on

Reply:
Yes, indeed. If we want to generate the entries in chronological order, it's not possible to do it without accumulation :(

Reply:
Maybe this is something that could be changed at the `factomd` level? I think it makes a lot of sense to have an endpoint, which returns the entries in a given chain in chronological order.